### PR TITLE
Possible solution for reporting attributes from Executioners

### PR DESCRIPTION
**Pull request by aeslaughter**
Here is a fairly generic method for reporting values from executioners, I implemented in a manner that may be expandable to other types of objects in the future if we deem it appropriate.

The first commit was my Eigen value specific solution that I was planning to use until I got into Yak and determined that there were multiple parameters being reported. I would like to leave this commit so I can come back to it if the more general method doesn't work out. 

The second commit creates a more generic Executioner reporting system, hopefully this or something similar will work, it will make cleaning up Yak much easier.

### CODE REVIEW

**Comment by aeslaughter** [comment]
closes #5037 
refs #4606

File: framework/src/postprocessors/ExecutionerAttributeReporter.C [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ExecutionerAttributeReporter::threadJoin(const UserObject & uo)
+{
+  const ExecutionerAttributeReporter & pps = static_cast<const ExecutionerAttributeReporter &>(uo);
+  if (_aggregation == 1)
+    _return_value += pps._return_value;
+  else if (_aggregation == 2)
+    _return_value = std::min(_return_value, pps._return_value);
+  else if (_aggregation == 3)
+    _return_value = std::max(_return_value, pps._return_value);
+}
```
Comment (permcody) [inline framework/src/postprocessors/ExecutionerAttributeReporter.C]: Don't need threadJoin in a GeneralPostprocessor

File: framework/src/executioners/EigenExecutionerBase.C [hunk 2]
```diff
@@ -233,7 +221,7 @@ EigenExecutionerBase::inversePowerIteration(unsigned int min_iter,
 
   // some iteration variables
   Real k_old = 0.0;
-  Real & source_integral_old = getPostprocessorValueOld("bx_norm");
+  Real source_integral_old = getPostprocessorValueOld("bx_norm");
```
Comment (permcody) [inline framework/src/executioners/EigenExecutionerBase.C]: Was this switch to a copy intentional?

**Comment by YaqiWang** [comment]
Does aggregation none means only value on master processor will be outputted? When can we call the addReporter?

**Comment by friedmud** [comment]
I don't support this.  I don't think we need anything to do with aggregation.

This thing should be REALLY simple.

**Comment by aeslaughter** [comment]
@friedmud What do you suggest then? This is very close to what was already being used in the ProblemRealParamater postprocessor. I added the aggregation piece so that future Executioner creators won't use this for reporting values that may be different across processors.

**Comment by permcody** [comment]
I think that's all that @friedmud is pointing out, we don't even have a case where we need aggregation right? @YaqiWang, in your executioner are you expecting your reportable value to vary across processors? Can you give us an example of where you'd need aggregation if you can think of one? Derek and I were thinking that anything that you might want to report (like your Eigenvalue) should already be consistent across processors by the time it gets reported.

**Comment by aeslaughter** [comment]
I don't think there is a case currently, I am just assuming that someone will see this and create some crazy executioner and wonder why things change in parallel.

This might all be a moot point, I just found some tests in Yak that use the "eigenvalue" parameter in Problem without an EigenExecutionerBase. I am likely going to need a more generic way to store a the Eigen value, probably with an UserObject. This whole Eigen thing is really a pain in the ...

**Comment by YaqiWang** [comment]
@permcody No, the eigenvalue I was reporting has been properly reduced across processors. @aeslaughter I know you want to make this general and it is not hard. But more functionality means more effort to maintain them, like documentation, tests, etc. We can always extend this when we found more use cases. For now, we can just add the reporter on master processor and thread. What I did before creating copies for each processor, which is not very desired.

**Comment by friedmud** [comment]
My point is that we don't need some complicated and code heavy system for this because we're not going to be able to guess all of the ways that people need to aggregate values... so we just shouldn't try.

If users need aggregation they can easily just do that themselves before reporting the value.

I really don't see what's so special about the Eigenvaue or why you would need a UserObject for it.

Personally (in my admittedly very simple application) I just computed the Eigenvalue _as_ a Postprocessor ( https://github.com/friedmud/gardensnake/blob/master/src/postprocessors/KEigenvalue.C ).  It's a perfect match for that because it really is a single scalar value reduction using integrals.  Now, in my case I did the integration in two other Postprocessors and then I coupled those into a GeneralPostprocessor to compute the Eigenvalue.  I did that because the other two number (fission sources) are useful as well (I was using them as convergence criteria).

It fit pretty easily in the MOOSE system really.

BUT: Rattlesnake is MUCH more complicated so I can see how he might want to do all of that in a custom Executioner.  In which case, having a simple ability to report a scalar value from an Executioner (that is already pre-aggregated) could be handy...

**Comment by aeslaughter** [comment]
OK, I will simplify the Executioner postprocessor stuff as much as possible.

The problem I am having with YAK is the various objects set/get a parameter named "eigenvalue" on FEProblem, I thought the actual usage of this parameter was limited to simulations that used EigenExecutionerBase based Executioners, so I moved the storage of the eigen value to the executioner and did dynamic casting to get/set the value via an accessor. However, there is a whole set of problems in YAK that rely on the "eigenvalue" parameter that do not use EigenExecutionerBase based executioners so my design is flawed.

I am thinking of just adding an `eigenValue()` method to FEProblem and then I can mimic the parameter based approach almost exactly.

**Comment by friedmud** [comment]
I see: I really think the Eigenvalue should be stored as a Postprocessor value.  Then objects can get it easily.  Just look at one of my objects here: https://github.com/friedmud/gardensnake/blob/master/src/kernels/Fission.C

"k" is the Eigenvalue Postprocessor.  I compute it as a Postprocessor... and then everything else can retrieve it and use it.

So this "reporting" system should just put the value in a Postprocessor and then everything else can retrieve it like normal.

**Comment by YaqiWang** [comment]
I think the correct way for storing eigenvalue should be a scalar variable, either primal or aux. EigenKernels should couple it. I did not implemented in the way because the eigenvalue is evaluated from a field variable like a postprocessor, and currently we can only do this with postprocessor. I think postprocessor should only be `general`. All the reset of postprocessors should be converted into scalar variables. But that will be a nightmare and I cannot find a way of gradually migrating to it.

**Comment by aeslaughter** [comment]
I like @friedmud's idea to make the Eigen value a postprocessor and have all values that need to use it get it like any other postprocessor.  This postprocessor could have a set method so that objects that need to change the value could do a dynamic cast on the Postprocessor and set the actual value. Does this seem like a reasonable solution?

Currently, there about about 15 different objects in YAK that set/get the "eigenvalue" parameter in FEProblem.

**Comment by friedmud** [comment]
I really think that even the computation of the Eigenvalue should get moved to Postprocessors.  That way you could trade out how it's computed easily.

However, that would take quite a bit of work and be of marginal benefit.  For now just having a way of setting the Postprocessor value from an Executioner will suffice.

At least that move will generalize all the objects that rely on the Eigenvalue to be relying on a Postprocessor... so then you COULD theoretically compute the Eigenvalue in a Postprocessor and have everything work at some point in the future.

It's a good first step anyway ;-)

**Comment by YaqiWang** [comment]
My previous comment may not be constructing.

If you look at EigenKernel constructor:
```
{
  if (_eigen)
  {
    _eigen_sys = static_cast<EigenSystem *>(&_fe_problem.getNonlinearSystem());
    _eigen_pp = _fe_problem.parameters().get<PostprocessorName>("eigen_postprocessor");
    if (_is_implicit)
      _eigenvalue = &getPostprocessorValueByName(_eigen_pp);
    else
      _eigenvalue = &getPostprocessorValueOldByName(_eigen_pp);
  }
  else
  {
    if (!_fe_problem.parameters().isParamValid("eigenvalue"))
      _fe_problem.parameters().set<Real>("eigenvalue") = 1.0;
    _eigenvalue = &_fe_problem.parameters().get<Real>("eigenvalue");
  }
}
```
you see `EigenKernel` needs a postprocessor storing eigenvalue even it is acting as a normal kernel. A default postprocessor needs to be created for eigen kernels, and its value can be set by Transfer for NDA stuff in Yak.

**Comment by friedmud** [comment]
@yaqiwang also mentioned another good option: make it a ScalarVariable.  That would have the nice side effect that you could leave it IN the nonlinear system and solve for it as part of your nonlinear solve (which is possible: but isn't used much from what I understand).

It's a toss up for me between ScalarVariable and Postprocessor.  I'll leave it up to @yaqiwang to decide which one makes more sense...

**Comment by YaqiWang** [comment]
This is a little off the topic.
Considering the eigenvalue problem:
`Ax = \lambda Bx`

Currently we can evaluate residual `R = \lambda Bx - Ax` when EigenKernels are indeed eigen and implicit. If we want to evaluate `Bx`, we have to copy the current solution to old and set the current solution to zero and make all eigen kernels explicit (i.e. make the implicit version of eigen kernels inactive, explicit version active), then we do residual evaluation to obtain `\lambda Bx`.

This is overall fine, maybe just some unnecessary vector copying is required. I am not sure how this will impact the possible SLEPSc implementation.

I prefer ScalarVariable for eigenvalue. I am willing to compromise with the current postprocessor way if @aeslaughter thinks the other way is too painful to do.

**Comment by aeslaughter** [comment]
I am happy to do whatever it takes to get this issue resolved, YAK is the
main thing holding me up on the constructor update.

To make the Eigen value a scalar we would need to create an AuxKernel to
perform the calculation, @YaqiWang can you do this portion?

**Comment by friedmud** [comment]
No: just make your reporter system set that value in the Aux solution vector.  Nothing else has to change...

**Comment by aeslaughter** [comment]
Just talked to @yaqiwang, he would perfer the postprocessor route. At least
for now.

**Comment by aeslaughter** [comment]
I am closing this for now.

**Pull request closed without merging**